File: fabric/src/main/java/me/lunaluna/fabric/elytrarecast/ElytraHelper.java
package me.lunaluna.fabric.elytrarecast;

import net.minecraft.client.network.ClientPlayerEntity;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.item.Items;
import net.minecraft.network.packet.c2s.play.ClientCommandC2SPacket;

public class ElytraHelper {
    public static boolean castElytra(ClientPlayerEntity player) {
        if (checkElytra(player) && checkFallFlyingIgnoreGround(player)) {
            var networkHandler = player.networkHandler;
            if (networkHandler != null)
                networkHandler
                        .sendPacket(new ClientCommandC2SPacket(player, ClientCommandC2SPacket.Mode.START_FALL_FLYING));
            return true;
        } else
            return false;
    }

    public static boolean checkElytra(ClientPlayerEntity player) {
        if (player.input.playerInput.jump() && !player.getAbilities().flying && !player.hasVehicle() && !player.isClimbing()) {
            var itemStack = player.getEquippedStack(EquipmentSlot.CHEST);
            return itemStack.isOf(Items.ELYTRA) && LivingEntity.canGlideWith(itemStack, EquipmentSlot.CHEST);
        } else
            return false;
    }

    public static boolean checkFallFlyingIgnoreGround(ClientPlayerEntity player) {
        if (!player.isTouchingWater() && !player.hasStatusEffect(StatusEffects.LEVITATION)) {
            var itemStack = player.getEquippedStack(EquipmentSlot.CHEST);
            if (itemStack.isOf(Items.ELYTRA) && LivingEntity.canGlideWith(itemStack, EquipmentSlot.CHEST)) {
                player.startGliding();
                return true;
            } else
                return false;
        } else
            return false;
    }
}

File: src/main/java/me/lunaluna/fabric/elytrarecast/ElytraHelper.kt
package me.lunaluna.fabric.elytrarecast

import net.minecraft.client.network.ClientPlayerEntity
import net.minecraft.entity.EquipmentSlot
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.item.ElytraItem
import net.minecraft.item.Items
import net.minecraft.network.packet.c2s.play.ClientCommandC2SPacket

object ElytraHelper {

    fun castElytra(player: ClientPlayerEntity) = if (checkElytra(player) && checkFallFlyingIgnoreGround(player)) {
        player.networkHandler?.sendPacket(
            ClientCommandC2SPacket(player, ClientCommandC2SPacket.Mode.START_FALL_FLYING)
        )
        true
    } else false

    private fun checkElytra(player: ClientPlayerEntity) = if (player.input.jumping && !player.abilities.flying && !player.hasVehicle() && !player.isClimbing) {
        val itemStack = player.getEquippedStack(EquipmentSlot.CHEST)
        itemStack.isOf(Items.ELYTRA) && ElytraItem.isUsable(itemStack)
    } else false

    private fun checkFallFlyingIgnoreGround(player: ClientPlayerEntity) = if (!player.isTouchingWater && !player.hasStatusEffect(StatusEffects.LEVITATION)) {
        val itemStack = player.getEquippedStack(EquipmentSlot.CHEST)
        if (itemStack.isOf(Items.ELYTRA) && ElytraItem.isUsable(itemStack)) {
            player.startFallFlying()
            true
        } else false
    } else false
}